tools/codegen/dest/lazy_ir.py

```diff
@@ -0,0 +1,387 @@
+    delegating the actual type conversion to `process_ir_types`
+    """
+    new_args_fields = {}
+    for arg_field in ["pre_self_positional",
+                      "self_arg",
+                      "post_self_positional",
+                      "pre_tensor_options_kwarg_only",
+                      "tensor_options",
+                      "post_tensor_options_kwarg_only",
+                      "out"]:
+        if arg_field == "self_arg" and getattr(func.arguments, "self_arg") is not None:
+            arg = getattr(func.arguments, "self_arg").argument
+            new_args_fields[arg_field] = SelfArgument(Argument(arg.name, process_ir_type(
+                arg.type, arg.is_write, arg.name), arg.default, arg.annotation))
+        elif getattr(func.arguments, arg_field) is not None:
+            new_args_fields[arg_field] = [
+                Argument(
+                    arg.name,
+                    process_ir_type(arg.type, arg.is_write, arg.name), arg.default, arg.annotation) for arg in getattr(func.arguments, arg_field)]
+        else:
+            new_args_fields[arg_field] = None
+    new_args = Arguments(**new_args_fields)
+    return FunctionSchema(func.name, new_args, func.returns)
+
+
+def separate_value_scalar_types(func: FunctionSchema) -> Tuple[List[NamedCType], List[NamedCType], List[NamedCType]]:
+    """
+    Since there are significant differences in how Values and 'everything else' are handled in the IR node class,
+    it's useful to have a way to get a list of just the values, just the scalars, or all of them.
+    """
+    types = [arg.type for arg in func.schema_order_arguments()]
+    value_types = [t for t in types if isValueType(t)]
+    scalar_types = [t for t in types if not isValueType(t)]
+    return types, value_types, scalar_types
+
+
+def node_ctor_inputs(value_types: List[NamedCType], scalar_types: List[NamedCType]) -> str:
+    """
+    Produce a formatted string with the arguments as passed into the constructor of a node class.
+    """
+    node_ctor_values = []
+    for t in value_types:
+        if isinstance(t.type, BaseCType):
+            node_ctor_values.append(f"l_{t.name}.GetIrValue()")
+        elif isinstance(t.type, OptionalCType):
+            node_ctor_values.append(
+                f"l_{t.name}.has_value() ? l_{t.name}.value().GetIrValue() : torch_lazy_tensors::ir::ops::kNullValue")
+        else:
+            assert False, ""
+
+    node_ctor_scalars = []
+    for t in scalar_types:
+        if isinstance(t.type, BaseCType) and t.type.type.name == "vector<int64_t>":
+            node_ctor_scalars.append(f"std::vector<int64_t>({t.name}.begin(), {t.name}.end())")
+        else:
+            node_ctor_scalars.append(t.name)
+
+    node_ctor_inputs = ",\n                              ".join(node_ctor_values + node_ctor_scalars)
+    return node_ctor_inputs
+
+
+def ir_node_name(func: FunctionSchema):
```

Comment:
Okay, I know we talked about this a little before - now I am wondering if there should just be an `api/lazy.py` file that contains most of these utility function mappings from `FunctionSchema` objects to various C++ names/strings/types.

If you look at `api/dispatcher.py`, it has several functions like:
`def name(func: FunctionSchema) -> str`
`def argument_type(a: Argument, *, binds: ArgName) -> NamedCType`
`def returns_type(rs: Sequence[Return]) -> CType`

each of which do some dispatcher-specific mapping, but fall back to the same api's in `api.cpp.py` in most cases. That sounds pretty similar to what most of these utility functions do. @ezyang